### ISSUE TITLE
Zero to Hero

### ISSUE THREAD

**Issue by wss020** [wss020/learning-log#1]

**Comment by wss020** [comment]
# 🧠 Neural Networks: Zero to Hero 学习笔记

> **Status:** 🚧 In Progress / ✅ Completed
> **Start Date:** 2026-02-18
> **Source:** [YouTube Playlist](https://www.youtube.com/watch?v=VMj-3S1tku0&list=PLAqhIrjkxbuWI23v9cThsA9GvCAUhRvKZ)
> **Author:** Andrej Karpathy

## 📖 学习目标
- **从零实现**：不依赖现成框架，手动实现 Backprop、BatchNorm、Attention 等核心机制。
- **理解本质**：通过手写 Micrograd 和 Makemore，彻底理解 PyTorch 也就是 `nn.Module` 背后的 Autograd 和 Tensor 运算原理。
- **复现 GPT**：最终能够手写一个简易版的 GPT transformer 模型。
---

## 📺 视频分集笔记 (Checklist)

### 1. The spelled-out intro to neural networks and backpropagation: building micrograd
- [x] **视频状态**: ⬜ 待看 / ⏳ 进行中 / ✅ 已完成
- **start date**:2026-02-18
- **end date**: 2026-02-24(偷大懒)
- **核心任务**:
  - 实现 `Value` 类（标量 autograd 引擎）。
  - 手动推导并实现加减乘除及 ReLU/Tanh 的 `.backward()`。
  - 用 Graphviz 可视化计算图。
  - 训练一个简单的 MLP 分类器。
- **关键知识点**:
  - Computational Graph (计算图)
  - Topological Sort (拓扑排序在反向传播中的作用)
  - Derivative / Chain Rule (链式法则的物理意义：局部梯度 * 上游梯度)
### 2. The spelled-out intro to language modeling: building makemore
- [ ] **视频状态**: ⬜ 待看 / ⏳ 进行中 / ✅ 已完成
- **核心任务**:
  - [ ] 实现 Bigram 字符级语言模型。
  - [ ] 引入 `torch.Tensor` 并理解 Broadcasting 机制。
  - [ ] 比较 "Counting" 方法与 "Neural Network" 方法的等价性。
  - [ ] 实现 Negative Log Likelihood (NLL) Loss。
- **关键知识点**:
  - Maximum Likelihood Estimation (MLE)
  - `torch.multinomial` 采样
  - One-hot encoding vs Embedding
- **我的笔记**:
### 3. Building makemore Part 2: MLP
- [ ] **视频状态**: ⬜ 待看 / ⏳ 进行中 / ✅ 已完成
- **核心任务**:
  - [ ] 复现 Bengio et al. 2003 论文 (Neural Probabilistic Language Models)。
  - [ ] 实现 Embedding Layer + Hidden Layer + Softmax。
  - [ ] 调整超参数 (Learning Rate, Hidden Size, Batch Size)。
- **关键知识点**:
  - Train / Dev / Test Split 的重要性
  - Overfitting vs Underfitting
  - Context Window (Block size) 的概念
- **我的笔记**:
  ### 4. Building makemore Part 3: Activations & Gradients, BatchNorm
- [ ] **视频状态**: ⬜ 待看 / ⏳ 进行中 / ✅ 已完成
- **核心任务**:
  - [ ] 解决梯度消失/爆炸问题 (Vanishing/Exploding Gradients)。
  - [ ] 手动实现 Kaiming Initialization。
  - [ ] **手写 Batch Normalization 层** (包含训练和推理时的 running mean/std)。
- **关键知识点**:
  - Activation Statistics (可视化激活值的分布)
  - Dead Neurons (ReLU 坏死问题)
  - BatchNorm 的 folding (推理时融合)
- **我的笔记**:
  ### 5. Building makemore Part 4: Becoming a Backprop Ninja
- [ ] **视频状态**: ⬜ 待看 / ⏳ 进行中 / ✅ 已完成
- **核心任务**:
  - [ ] **不使用 Autograd**，手动推导并手写 MLP（含 BatchNorm/CrossEntropy）的每一层反向传播。
  - [ ] 能够通过 `torch.autograd.gradcheck`。
- **关键知识点**:
  - Matrix Calculus (矩阵微积分：形状检查大法)
  - Cross Entropy Loss 的梯度推导简洁形式 (`softmax - target`)
  - Broadcasting 在反向传播中的处理 (Sum)
- **我的笔记**:
  ### 6. Building makemore Part 5: Building a WaveNet
- [ ] **视频状态**: ⬜ 待看 / ⏳ 进行中 / ✅ 已完成
- **核心任务**:
  - [ ] 将 MLP 升级为类似 CNN 的结构 (Dilated Convolutions 概念)。
  - [ ] 优化代码结构，使其更像 PyTorch 的 `nn.Sequential`。
- **关键知识点**:
  - Stride & Dilation
  - Tree-like hierarchical structure
  - 代码重构与模块化设计
- **我的笔记**:
  ### 7. Let's build GPT: from scratch, in code, spelled out
- [ ] **视频状态**: ⬜ 待看 / ⏳ 进行中 / ✅ 已完成
- **核心任务**:
  - [ ] 彻底理解 Self-Attention 机制。
  - [ ] 实现 Multi-Head Attention。
  - [ ] 加上 Feed-Forward Network 和 Residual Connections (残差连接)。
  - [ ] 加上 LayerNorm 和 Dropout。
  - [ ] 训练一个莎士比亚风格的 GPT 模型。
- **关键知识点**:
  - Query, Key, Value 的直观理解
  - Masked Attention (Decoder-only 架构)
  - Positional Embeddings
  - Scaled Dot-Product Attention
- **我的笔记**:
  ### 8. Let's build the GPT Tokenizer
- [ ] **视频状态**: ⬜ 待看 / ⏳ 进行中 / ✅ 已完成
- **核心任务**:
  - [ ] 理解 Unicode (UTF-8) 编码。
  - [ ] 实现 Byte Pair Encoding (BPE) 算法。
  - [ ] 处理 Regex split pattern (GPT-2/GPT-4 的特殊处理)。
- **关键知识点**:
  - Tokenization 导致的常见 LLM 问题 (算术、拼写、代码缩进)
  - Vocabulary Size trade-offs
- **我的笔记**:
  ### (Optional) 9. Let's reproduce GPT-2 (124M)
- [ ] **视频状态**: ⬜ 待看 / ⏳ 进行中 / ✅ 已完成
- **说明**: 这是 Karpathy 最新的超硬核直播，时长 4 小时，从零复现 GPT-2 124M 版本并在 C 语言 (llm.c) 中优化。
- **我的笔记**:
  ---

## 💡 核心收获总结 (Summary)

### 1. 关于 Autograd
> *在此总结：为什么我们需要 Autograd？手动实现一遍 backward 之后，你对 PyTorch 的 `.backward()` 有了什么新的认识？*

### 2. 关于 Tensor 与 Shape
> *在此总结：在处理高维数据（Batch, Time, Channel）时，你总结出了哪些调试 Shape 的技巧？*

### 3. 关于 Transformer & Attention
> *在此总结：为什么 Attention 机制如此有效？它是如何解决长距离依赖问题的？*

---

## 🔗 参考资料
- [Karpathy's GitHub - micrograd](https://github.com/karpathy/micrograd)
- [Karpathy's GitHub - makemore](https://github.com/karpathy/makemore)
- [Karpathy's GitHub - nanoGPT](https://github.com/karpathy/nanoGPT)
- [Attention Is All You Need Paper](https://arxiv.org/abs/1706.03762)

**Comment by wss020** [comment]
### 1. The spelled-out intro to neural networks and backpropagation: building micrograd
Micrograd
前面50min带着简单manual back propagation（怎么导数、链式法则）
>a recursive applicaiton of chain rule backwards through the computation graph

1:18:00 Assign the _backward closure to the out object to propagate gradients back to self and other.

>[!NOTE]
>__闭包 (Closure)__
>在 __mul__ 等操作中定义的 def _backward(): 是一个闭包 (Closure)。
> · 作用：它像一个“时间胶囊”，捕获并锁定了前向传播时的上下文（即参与运算的 self 和 other 节点）。
> · 机制：虽然 __mul__ 执行结束了，但 _backward 依然保留了对 self 和 other 的引用。
> · 目的：这使得我们在后续调用 loss.backward() 时，能够访问当初的数据 (self.data) 来计算导数，而不需要手动创建一个 >Context 对象来存储这些中间值。

当一个parameter用了多次，gradients相加，so __backward()中都改成+=
>[!NOTE]
>$$f'(x) = \frac{d}{dx}(e^x) = e^x$$
